File: mobile/lib/protocol/pong_payload.dart
import 'dart:typed_data';

import 'payload_codec.dart';

class PongPayload {
  const PongPayload({required this.clientSendTime, required this.serverTime});

  final int clientSendTime;
  final int serverTime;

  @override
  bool operator ==(Object other) =>
      other is PongPayload &&
      clientSendTime == other.clientSendTime &&
      serverTime == other.serverTime;

  @override
  int get hashCode => Object.hash(clientSendTime, serverTime);
}

class PongPayloadCodec {
  static Uint8List encode(PongPayload payload) {
    final writer = PayloadWriter()
      ..writeUInt64(payload.clientSendTime)
      ..writeUInt64(payload.serverTime);
    return writer.toBytes();
  }

  static PongPayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final clientSendTime = reader.readUInt64();
    final serverTime = reader.readUInt64();
    reader.expectEnd();
    return PongPayload(clientSendTime: clientSendTime, serverTime: serverTime);
  }
}

File: desktop/Protocol/PongPayload.cs
namespace CTRL.Desktop.Protocol;

public sealed record PongPayload(ulong ClientSendTime, ulong ServerTime);

public static class PongPayloadCodec
{
    public static byte[] Encode(PongPayload payload)
    {
        var writer = new PayloadWriter();
        writer.WriteUInt64(payload.ClientSendTime);
        writer.WriteUInt64(payload.ServerTime);
        return writer.ToArray();
    }

    public static PongPayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var clientSendTime = reader.ReadUInt64();
        var serverTime = reader.ReadUInt64();
        reader.ExpectEnd();
        return new PongPayload(clientSendTime, serverTime);
    }
}
